### PR TITLE
First stab at caching more RAR info

### PR DESCRIPTION
**Pull request by Forgind**
This gives RAR the ability to save off a modified state file and load it for later use.

@rainersigwald, would you mind looking at this at some point to let me know if you think it's going in the direction you expected? This is certainly incomplete—proposed future steps below.

Note: As far as I can tell, the only file path stored and serialized (currently) is the key—that is, FileStates didn't have any paths in them—but the sample solutions I tried didn't have scatter files, so I was unsure about those. As far as I could tell, they're never used? That felt wrong, but it's internal, and there were very few references to it, and they seemed to be assigning to it with one exception that could be using it in a very indirect way.

Still to do:
 - [x] Verify that the precomputed cache read in is real.
 - [x] Change logging such that it will log messages instead of errors or warnings if there are problems with deserializing the precomputed cache
 - [x] Comments at relevant points
 - [x] Switch serialization mode to another form (json?)
 - [x] Validation + performance tests
 - [ ] Have SDK opt in

### CODE REVIEW

File: src/Tasks/SystemState.cs [hunk 1]
```diff
@@ -547,13 +552,58 @@ out fileState.frameworkName
...
+                foreach (string s in sfBase.instanceLocalFileStateCache.Keys)
+                {
+                    if (!retVal.instanceLocalFileStateCache.ContainsKey(s))
+                    {
+                        FileState fileState = (FileState)sfBase.instanceLocalFileStateCache[s];
+                        // Verify that the assembly is correct
+                        if (File.Exists(s) && Assembly.LoadFile(s).ManifestModule.ModuleVersionId.Equals(fileState.ModuleVersionID))
+                        {
+                            // Correct file path and timestamp
+                            string fullPath = Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), stateFile));
```
Comment (rainersigwald) [inline src/Tasks/SystemState.cs:575]: directory of stateFile, not current directory, right?

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs [hunk 2]
```diff
@@ -1845,23 +1849,36 @@ private void LogConflict(Reference reference, string fusionName)
         /// <summary>
         /// Reads the state file (if present) into the cache.
         /// </summary>
-        private void ReadStateFile()
+        private void ReadStateFile(GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo)
```
Comment (rainersigwald) [inline src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1852]: Why this change?

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs [hunk 3]
```diff
@@ -441,6 +441,10 @@ public string TargetedRuntimeVersion
             set { _targetedRuntimeVersionRawValue = value; }
         }
 
+        public string PreComputedCacheOutputPath { get; set; }
```
Comment (rainersigwald) [inline src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:444]: I don't like this much better, would like to think on a name for these.

```suggestion
        public string OutputCache { get; set; }
```

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs [hunk 4]
```diff
@@ -441,6 +441,10 @@ public string TargetedRuntimeVersion
...
+
+        public string[] PreComputedCacheFileList { get; set; }
```
Comment (rainersigwald) [inline src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:446]:
```suggestion
        public string[] AssemblyInformationCache { get; set; }
```

?

File: src/Tasks/SystemState.cs [hunk 5]
```diff
@@ -547,13 +552,58 @@ out fileState.frameworkName
...
+
+            return retVal;
+        }
+
+        internal void SerializePrecomputedCache(string stateFile, TaskLoggingHelper log)
+        {
+            foreach (string path in instanceLocalFileStateCache.Keys)
+            {
+                FileState fileState = (FileState)instanceLocalFileStateCache[path];
+                fileState.ModuleVersionID = Assembly.LoadFrom(path).ManifestModule.ModuleVersionId;
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks/SystemState.cs:591]: Does `Assembly.ReflectionOnlyLoadFrom(path)` also work? It seems clearer to me.

File: src/Tasks/SystemState.cs (hunk 1 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:575]: Yes—directory of stateFile, and the relative path is s. (I should probably change that name, too.)

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs (hunk 2 above)
Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1852]: I know you said GetLastWriteTime was only provided for testing, but this is in case that ever changes. I wanted to update the `LastModified` field of each fileState to what it would be if it's queried later, and since it technically uses getLastWriteTime, I wanted to use the appropriate getLastWriteTime, which meant passing it through to DeserializePrecomputedCaches and using it to adjust LastModified rather than using the real timestamp.

Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1852]: I added installedAssemblyTableInfo because they were set together, and it seemed wrong to split them. That part wasn't actually needed, though.

File: src/Tasks/SystemState.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Tasks/SystemState.cs:591]: It does not. That also locks the assembly.

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs (hunk 3 above)
Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:444]: How about CacheOutputPath, Relative(Path)CacheOutputPath, or OutputCachePath?

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs (hunk 4 above)
Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:446]: "AssemblyInformation" just sounds like "related to RAR" to me. It would probably be best to match this with the above name, so the equivalents of my latest suggestions would be (Secondary)CacheInputPath, Relative(Path)CacheInputPath, or InputCachePath. Do any of those sound reasonable to you?

**Comment by wli3** [comment]
Don't block on my approval. I am not very familiar with MSBuild. Although I will read this

File: src/Tasks/SystemState.cs [hunk 6]
```diff
@@ -192,22 +196,24 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
...
                 }
+                info.AddValue("mvid", ModuleVersionID);
             }
 
             /// <summary>
             /// Gets the last modified date.
             /// </summary>
             /// <value></value>
-            internal DateTime LastModified
+            public DateTime LastModified
```
Comment (Forgind) [inline src/Tasks/SystemState.cs:206]: These were made public as part of trying to switch to json. It seemed to be serializing/deserializing only public properties in my testing console even if it had access, but I think I'm still missing a piece.

**Comment by BenVillalobos** [comment]
PR nit: 

- [ ] Verify that the precomputed cache read in is real.
- [ ] Change logging such that it will log messages instead of errors or warnings if there are problems with deserializing the precomputed cache
- [ ] Comments at relevant points
- [ ] Switch serialization mode to another form (json?)
- [ ] Validation + performance tests
- [ ] Have SDK opt in

Checkboxes! 😉

**Comment by wli3** [comment]
Looks like we are still in explore phase, trying out it will help the perf or not. But once we have are sure about this route, we should have unit test coverage.

File: src/Tasks/SystemState.cs [hunk 7]
```diff
@@ -115,7 +118,7 @@ private sealed class FileState : ISerializable
             /// <summary>
             /// The last modified time for this file.
             /// </summary>
-            private DateTime lastModified;
+            internal DateTime lastModified;
```
Comment (ladipro) [inline src/Tasks/SystemState.cs:121]: nit: This change shouldn't be needed as JSON serialization ignores fields.

File: src/Tasks/SystemState.cs [hunk 8]
```diff
@@ -548,12 +559,85 @@ out fileState.frameworkName
...
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Modifies this object to be more portable across machines, then writes it to stateFile.
+        /// </summary>
+        internal void SerializePrecomputedCache(string stateFile, TaskLoggingHelper log)
+        {
+            Hashtable newInstanceLocalFileStateCache = new Hashtable();
```
Comment (ladipro) [inline src/Tasks/SystemState.cs:611]:
```suggestion
            Hashtable newInstanceLocalFileStateCache = new Hashtable(instanceLocalFileStateCache.Count);
```

File: src/Tasks/SystemState.cs [hunk 9]
```diff
@@ -548,12 +559,85 @@ out fileState.frameworkName
...
+            foreach (string path in instanceLocalFileStateCache.Keys)
+            {
+                // Add MVID to allow us to verify that we are using the same assembly later
+                FileState fileState = (FileState)instanceLocalFileStateCache[path];
```
Comment (ladipro) [inline src/Tasks/SystemState.cs:615]: nit: Enumerate `instanceLocalFileStateCache` and not `instanceLocalFileStateCache.Keys` to avoid another hashtable lookup here.

**Review by ladipro — COMMENTED** [review]
Have you had a chance to measure how long it takes to read an assembly MVID versus read & parse its references?

File: src/Tasks/StateFileBase.cs [hunk 10]
```diff
@@ -64,9 +65,10 @@ internal virtual void SerializeCache(string stateFile, TaskLoggingHelper log)
         /// <summary>
         /// Reads the specified file from disk into a StateFileBase derived object.
         /// </summary>
-        internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelper log, Type requiredReturnType)
+        internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelper log, Type requiredReturnType, bool logWarnings = true)
```
Comment (cdmihai) [inline src/Tasks/StateFileBase.cs:68]: Looking at the usages of this method, wouldn't it look nicer as a generic method?

File: src/Tasks/SystemState.cs [hunk 11]
```diff
@@ -548,12 +559,91 @@ out fileState.frameworkName
...
+
+            foreach (string stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sysState = (SystemState)DeserializeCache(stateFile, log, requiredReturnType, false);
+                if (sysState == null)
+                {
+                    continue;
+                }
+                foreach (DictionaryEntry kvp in sysState.instanceLocalFileStateCache)
```
Comment (cdmihai) [inline src/Tasks/SystemState.cs:580]: nit: make instanceLocalFileStateCache generic.

File: src/Tasks/SystemState.cs [hunk 12]
```diff
@@ -548,12 +559,91 @@ out fileState.frameworkName
...
+                        Guid mvid;
+                        string fullPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(stateFile), relativePath));
+                        if (File.Exists(fullPath))
+                        {
+                            using (var reader = new PEReader(File.OpenRead(fullPath)))
+                            {
+                                var metadataReader = reader.GetMetadataReader();
+                                mvid = metadataReader.GetGuid(metadataReader.GetModuleDefinition().Mvid);
+                            }
+                            if (mvid.Equals(fileState.ModuleVersionID))
```
Comment (cdmihai) [inline src/Tasks/SystemState.cs:596]: Why is the mvid check required? What can go wrong if the check didn't exist?

File: src/Tasks/SystemState.cs [hunk 13]
```diff
@@ -548,12 +559,91 @@ out fileState.frameworkName
...
+                foreach (DictionaryEntry kvp in sysState.instanceLocalFileStateCache)
+                {
+                    string relativePath = (string)kvp.Key;
+                    if (!assembliesFound.Contains(relativePath))
+                    {
+                        FileState fileState = (FileState)kvp.Value;
+                        // Verify that the assembly is correct
+                        Guid mvid;
+                        string fullPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(stateFile), relativePath));
+                        if (File.Exists(fullPath))
```
Comment (cdmihai) [inline src/Tasks/SystemState.cs:589]: use `FileSystems.Default`

File: src/Tasks/SystemState.cs (hunk 12 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:596]: This cache is expected to come from another machine, and you could have a different version of the assembly.

File: src/Tasks/SystemState.cs (hunk 11 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:580]: I don't understand. Do you mean to make the type of instanceLocalFileStateCache more generic, change DictionaryEntry to var, make it so this can pull from something other than a system state by making the input to this loop instanceLocalFileStateCache instead of sysState, or something else?

File: src/Tasks/SystemState.cs (hunk 12 above)
Comment (cdmihai) [reply src/Tasks/SystemState.cs:596]: I see, and the mvid is the only ID you found that's guaranteed to change, as people can lie about the assembly version (like we do). I wonder if using the PEReader to parse the assembly mvid is faster than just hashing the assembly contents as a blob.

File: src/Tasks/SystemState.cs (hunk 11 above)
Comment (cdmihai) [reply src/Tasks/SystemState.cs:580]: Make it a generic dictionary.

File: src/Tasks/SystemState.cs (hunk 12 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:596]: Right. I imagine which is faster would depend heavily on how big the assembly is—if it's tiny, you may as well hash the whole thing. If it's huge, mvids are the way to go.

Comment (cdmihai) [reply src/Tasks/SystemState.cs:597]: Yup. Curious what the median BCL assembly size is.

**Comment by Forgind** [comment]
Did my first performance test, comparing building MSBuild.sln (with no caches) to building MSBuild.sln with only the precomputed caches. Here are the times for RAR:
No caches:
image
Only precomputed caches:
image

Although there's always a nontrivial chance that I did something wrong, and these results are invalid, they look promising!

File: src/Tasks/SystemState.cs [hunk 14]
```diff
@@ -544,12 +556,90 @@ out fileState.frameworkName
...
+        }
+
+        /// <summary>
+        /// Modifies this object to be more portable across machines, then writes it to stateFile.
+        /// </summary>
+        internal void SerializePrecomputedCache(string stateFile, TaskLoggingHelper log, Func<string, Guid> calculateMvid)
+        {
+            Dictionary<string, FileState> newInstanceLocalFileStateCache = new Dictionary<string, FileState>(instanceLocalFileStateCache.Count);
+            calculateMvid ??= CalculateMvid;
+            foreach (KeyValuePair<string, FileState> kvp in instanceLocalFileStateCache)
```
Comment (BenVillalobos) [inline src/Tasks/SystemState.cs:608]: How large does the `instanceLocalFileStateCache` dictionary get on average?

Thinking out loud: I wonder what difference it would make to `foreach` over the keys and look up the values vs iterating over the kvp's.

File: src/Tasks/StateFileBase.cs [hunk 15]
```diff
@@ -77,32 +78,35 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
                     using (FileStream s = new FileStream(stateFile, FileMode.Open))
                     {
                         var formatter = new BinaryFormatter();
```
Comment (BenVillalobos) [inline src/Tasks/StateFileBase.cs:80]: The commit history shows you reverted back to a binaryformatter over json serialization. What was the reason for this again?

File: src/Tasks/SystemState.cs [hunk 16]
```diff
@@ -544,12 +556,90 @@ out fileState.frameworkName
...
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sysState = DeserializeCache<SystemState>(stateFile.ToString(), log, false);
+                if (sysState == null)
+                {
+                    continue;
+                }
+                foreach (KeyValuePair<string, FileState> kvp in sysState.instanceLocalFileStateCache)
+                {
+                    string relativePath = kvp.Key;
+                    if (!assembliesFound.Contains(relativePath))
```
Comment (BenVillalobos) [inline src/Tasks/SystemState.cs:582]: Is this check necessary? You're iterating through a dictionary, there will be no duplicate keys.

Comment (BenVillalobos) [reply src/Tasks/SystemState.cs:582]: Ah, I see there may be duplicates between files. Will order matter here? Will there be some future file that has more information than a previous one, and therefore that info won't get logged?

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+                    new TaskItem(precomputedCachePath)
+                };
+
+                u.ReadStateFile(File.GetLastWriteTime, Array.Empty<AssemblyTableInfo>(), calculateMvid, p => true);
+                u._cache.instanceLocalFileStateCache.ShouldNotContainKey(dllName);
+                File.Delete(standardCache.Path);
+                u._cache = null;
+                u.ReadStateFile(File.GetLastWriteTime, Array.Empty<AssemblyTableInfo>(), calculateMvid, p => true);
+                u._cache.instanceLocalFileStateCache.ShouldContainKey(dllName);
+                SystemState.FileState a3 = u._cache.instanceLocalFileStateCache[dllName];
```
Comment (BenVillalobos) [inline src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:95]: nit: `t`, `u`, and `a3` aren't great variable names.

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+                t._cache = new SystemState();
+                t._cache.instanceLocalFileStateCache = new Dictionary<string, SystemState.FileState>() {
+                    { Path.Combine(standardCache.Path, "assembly1"), new SystemState.FileState(DateTime.Now) },
+                    { Path.Combine(standardCache.Path, "assembly2"), new SystemState.FileState(DateTime.Now) { Assembly = new Shared.AssemblyNameExtension("hi") } } };
+                t._cache.IsDirty = true;
+                t.StateFile = standardCache.Path;
+                t.WriteStateFile(calculateMvid);
+                int standardLen = File.ReadAllText(standardCache.Path).Length;
+                File.Delete(standardCache.Path);
+                standardLen.ShouldBeGreaterThan(0);
```
Comment (BenVillalobos) [inline src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:43]: Is this a strong enough check? Should we directly compare `File.ReadAllText` to exactly what we think the output should be?

**Review by BenVillalobos — APPROVED** [review]
LGTM, my comments aren't merge-blocking.

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs (hunk 18 above)
Comment (Forgind) [reply src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:43]: I'd love to, but unfortunately, that output would differ depending on serialization method.

File: src/Tasks/SystemState.cs (hunk 16 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:582]: No because any particular key with a valid mvid will be identical to any identical key/mvid pair.

File: src/Tasks/StateFileBase.cs (hunk 15 above)
Comment (Forgind) [reply src/Tasks/StateFileBase.cs:80]: Had trouble getting it to work. Made it work in #5704.

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs (hunk 17 above)
Comment (cdmihai) [reply src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:95]: +1

---
In reply to: [505966901](https://github.com/dotnet/msbuild/pull/5514#discussion_r505966901) [](ancestors = 505966901)

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+                {
+                    new TaskItem(precomputedCachePath)
+                };
+
+                u.ReadStateFile(File.GetLastWriteTime, Array.Empty<AssemblyTableInfo>(), calculateMvid, p => true);
+                u._cache.instanceLocalFileStateCache.ShouldNotContainKey(dllName);
+                File.Delete(standardCache.Path);
+                u._cache = null;
+                u.ReadStateFile(File.GetLastWriteTime, Array.Empty<AssemblyTableInfo>(), calculateMvid, p => true);
+                u._cache.instanceLocalFileStateCache.ShouldContainKey(dllName);
```
Comment (cdmihai) [inline src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:94]: Leave a comment saying why the entry exists here but not at line 90.

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs [hunk 20]
```diff
@@ -1855,25 +1855,33 @@ private void LogConflict(Reference reference, string fusionName)
...
         /// </summary>
-        private void WriteStateFile()
+        internal void WriteStateFile(Func<string, Guid> calculateMvid = null)
         {
-            if (!string.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
+            if (!string.IsNullOrEmpty(AssemblyInformationCacheOutputPath))
+            {
+                _cache.SerializePrecomputedCache(AssemblyInformationCacheOutputPath, Log, calculateMvid);
+            }
+            else if (!string.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
```
Comment (cdmihai) [inline src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1884]: Does it make any sense to run the else branch even when `AssemblyInformationCacheOutputPath` is set?

Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1884]: I don't think there's any need to. It wouldn't be a bad idea to switch the order of these; the first is expected to almost never be true, and if it is, you're creating a cache to send to users, and since that would likely only happen on a remote build machine, you probably don't care about local state.

File: src/Tasks/SystemState.cs [hunk 21]
```diff
@@ -31,7 +35,7 @@ internal sealed class SystemState : StateFileBase, ISerializable
         /// <summary>
         /// Cache at the SystemState instance level. It is serialized and reused between instances.
         /// </summary>
-        private Hashtable instanceLocalFileStateCache = new Hashtable(StringComparer.OrdinalIgnoreCase);
+        internal Dictionary<string, FileState> instanceLocalFileStateCache = new Dictionary<string, FileState>(StringComparer.OrdinalIgnoreCase);
```
Comment (cdmihai) [inline src/Tasks/SystemState.cs:38]: Nice, nice.

File: src/Tasks/Resources/Strings.resx [hunk 22]
```diff
@@ -451,6 +451,10 @@
     <value>MSB3101: Could not write state file "{0}". {1}</value>
     <comment>{StrBegin="MSB3101: "}</comment>
   </data>
+  <data name="General.StateFileAlreadyPresent">
+    <value>MSB3101: There is already a file at "{0}". If you are trying to create a precomputed cache, ensure that you are building a single project that depends on your assemblies rather than building your assemblies themselves. If you are running RAR normally, do not set the "_CacheOutputPath" field of RAR.</value>
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply src/Tasks/Resources/Strings.resx]: Yea, make RAR explicit.

---
In reply to: [477323537](https://github.com/dotnet/msbuild/pull/5514#discussion_r477323537) [](ancestors = 477323537)

**Review by cdmihai — APPROVED** [review]

File: src/Tasks/SystemState.cs (hunk 16 above)
Comment (cdmihai) [inline src/Tasks/SystemState.cs:582]: What about conflicts? You could either error (as a sanity check that conflicts don't happen) or you could keep the newest one.

Comment (Forgind) [reply src/Tasks/SystemState.cs:582]: If there are conflicts, the first one was already verified, so I skipped checking later ones that might be invalid to save a little time. I could easily imagine collisions, but if they both have the right mvid, they should be identical.

Comment (cdmihai) [reply src/Tasks/SystemState.cs:582]: What if two files contain the same relative path but with different mvids?

**Comment by Forgind** [comment]
I build OrchardCore once with no caches of any kind and once with the perfect precomputed caches. Results:
NoCaches:
Mean: 85838.3843
Median: 82582.754
PrecomputedCaches:
Mean: 61150.18
Median: 58261.5

Looking at the full data, I noticed more powerfully a trend hinted at by the data with building MSBuild: precomputed caches make it a little slower if RAR would be fast either way but much faster if it would actually have to calculate something. This make sense, since deserializing and validating the cache takes some time. More precisely, having no caches made RAR faster for the fastest forty (out of 159) RAR executions, ranging up to about 30 seconds. After that, however, the time it took increased much faster, so whereas the difference between the fastest RAR executions for the two was about 100ms, the difference between having precomputed caches and not having them was over a minute for the slowest execution.

File: src/Tasks/SystemState.cs [hunk 23]
```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
...
+                {
+                    string relativePath = kvp.Key;
+                    if (!assembliesFound.Contains(relativePath))
+                    {
+                        FileState fileState = kvp.Value;
+                        string fullPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(stateFile.ToString()), relativePath));
+                        if (fileExists(fullPath))
+                        {
+                            // Correct file path and timestamp
+                            fileState.LastModified = retVal.getLastWriteTime(fullPath);
```
Comment (rainersigwald) [inline src/Tasks/SystemState.cs:581]: Is this the right timestamp? Doesn't it mean that we would potentially get incorrect information if the user updated one of the cached assemblies before a clean build?

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs (hunk 18 above)
Comment (rainersigwald) [reply src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:34]: I don't think I understand the purpose of this test. Isn't it a subset of the other one that does a round trip?

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+                    new TaskItem(precomputedCachePath)
+                };
+
+                // At this point, we should have created two cache files: one "normal" one and one "precomputed" one.
+                // When we read the state file the first time, it should read from the caches produced in a normal
+                // build, partially because we can read it faster. If that cache does not exist, as with the second
+                // time we try to read the state file, it defaults to reading the "precomputed" cache. In this case,
+                // the normal cache does not have dll.dll, whereas the precomputed cache does, so it should not be
+                // present when we read the first time but should be present the second time. Then we verify that the
+                // information contained in that cache matches what we'd expect.
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:88]: I find this very confusing. Can you split it into two tests:

1. "Normal project-level StateFile cache takes precedence over AssemblyInformationCachePaths", and
2. "Writing and then reading AssemblyInformationCachePaths preserves all relevant information"?

File: src/Tasks/SystemState.cs [hunk 25]
```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
...
+            {
+                string relativePath = FileUtilities.MakeRelative(Path.GetDirectoryName(stateFile), kvp.Key);
+                newInstanceLocalFileStateCache[relativePath] = kvp.Value;
+            }
+            instanceLocalFileStateCache = newInstanceLocalFileStateCache;
+
+            if (FileUtilities.FileExistsNoThrow(stateFile))
+            {
+                log.LogWarningWithCodeFromResources("General.StateFileAlreadyPresent", stateFile);
+            }
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks/SystemState.cs:608]: We decided this in a meeting a while back. Most people won't use this (_making_ a cache) at all, and those who do may be confused as to whether they should be building their solution to create the cache or building a project that uses all the assemblies from their solution to create the cache. This attempts to push for the correct behavior.

File: src/Tasks/SystemState.cs [hunk 26]
```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
...
+
+        /// <summary>
+        /// Modifies this object to be more portable across machines, then writes it to stateFile.
+        /// </summary>
+        internal void SerializePrecomputedCache(string stateFile, TaskLoggingHelper log)
+        {
+            Dictionary<string, FileState> newInstanceLocalFileStateCache = new Dictionary<string, FileState>(instanceLocalFileStateCache.Count);
+            foreach (KeyValuePair<string, FileState> kvp in instanceLocalFileStateCache)
+            {
+                string relativePath = FileUtilities.MakeRelative(Path.GetDirectoryName(stateFile), kvp.Key);
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks/SystemState.cs:600]: I believe so. The cache will be serialized to a particular location relative to the SDK assemblies on the computer that builds the SDK, and when the user downloads both those assemblies and this cache, they will be in the same place relative to each other.

File: src/Tasks/SystemState.cs [hunk 27]
```diff
@@ -282,6 +287,7 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
         internal bool IsDirty
         {
             get { return isDirty; }
+            set { isDirty = value; }
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks/SystemState.cs:290]: Yes

File: src/Tasks/SystemState.cs [hunk 28]
```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
...
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+            fileExists ??= FileSystems.Default.FileExists;
+
+            foreach (ITaskItem stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sysState = DeserializeCache<SystemState>(stateFile.ToString(), log, false);
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks/SystemState.cs:566]: I hadn't been thinking about that. The way I've been using them is passing in full paths, and users could get those from property + relative path if necessary. It's certainly limiting to make this assumption, but since it's new functionality, no one will be broken by it being restricted in that way, and it would certainly be faster to just know that it's a full path to the cache. I could go either way. It sounds like you'd prefer it account for globs and paths from other projects? We might need to talk about that if so, since I don't think I know all of how to do that.

File: src/Tasks/SystemState.cs [hunk 29]
```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
...
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+            fileExists ??= FileSystems.Default.FileExists;
+
+            foreach (ITaskItem stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Tasks/SystemState.cs:565]: I think it was a compat concern. Old state file uses serializationMethod_1, but user is using new version of MSBuild that expected seralizationMethod_2 --> we can't read the file, but we shouldn't error. It's better to have one slow build (then revert to normal speed) than to send an error and force the user to try to figure it out. This will hopefully be an edge case anyway, since a lot of the use of these caches is with new users.

File: src/Tasks/SystemState.cs (hunk 23 above)
Comment (Forgind) [reply src/Tasks/SystemState.cs:581]: Yes! Thank you for pointing that out. I removed that line entirely, and it would have made the cache (often) wrong. This was good behavior when I was checking for mvids but is no longer correct, and I didn't remove it.

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs (hunk 20 above)
Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1878]: I guess my question would be "why are you creating a precomputed cache if you're just doing an incremental build"? The precomputed caches are for sending to customers, and they're only useful when building from an otherwise cache-free state. Note that if I remember correctly, serializing caches took ~7% of RAR's total time, and serializing two caches would double the time spent on it.

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs (hunk 18 above)
Comment (Forgind) [reply src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:34]: Kinda, yeah. I still favor leaving it in because 1) it's already written, and 2) if there's a problem here (nontrivial possibility, since this is new), having a simpler test will more readily provide insight as to what exactly is broken.
